probe-rs-tools/src/bin/probe-rs/cmd/dap_server/debug_adapter/dap/request_helpers.rs

```diff
@@ -95,101 +96,147 @@ pub(crate) fn disassemble_target_memory(
+        // Find the closest source location to ensure that we're starting
+        // with a well-aligned instruction pointer. Note: Variable
+        // length instructions are not necessarily word-aligned, i.e.
+        // in the case of ARM Thumbv2, instructions are embedded into
+        // a 16-bit halfword stream.
+        if let Some(source_location) = target_core
+            .core_data
+            .debug_info
+            .get_source_location(start_from_address)
+        {
+            if let Some(source_address) = source_location.address {
+                start_from_address = source_address;
             }
         }
+    }
+
+    // Ensure pointer alignment (safety measure, should be a no-op).
+    start_from_address &= !(min_instruction_size - 1);
+    read_until_address &= !(min_instruction_size - 1);
+
+    let cs = get_capstone(target_core)?;
+    let mut code_buffer_le: Vec<u8> = vec![];
+    let mut disassembled_instructions: Vec<DisassembledInstruction> = vec![];
+    let mut maybe_previous_source_location = None;
+    let mut maybe_reference_instruction_index = None;
+
+    let mut instruction_pointer = start_from_address;
+    'instruction_loop: while instruction_pointer < read_until_address {
+        if maybe_reference_instruction_index.is_none()
+            && instruction_pointer >= adjusted_memory_reference
+        {
+            // This instruction will be the one that the requested memory
+            // reference points to. We'll calculate instruction offsets
+            // relative to this index.
+            maybe_reference_instruction_index = Some(disassembled_instructions.len() as i64);
+        }
+
+        let mut read_pointer = instruction_pointer + code_buffer_le.len() as u64;
+        let mut read_error = None;
+        while read_error.is_none() && code_buffer_le.len() < max_instruction_size as usize {
+            // Ensure that we buffer at least one full instruction.
+            let bytes_to_read = max_instruction_size - code_buffer_le.len() as u64;
+            let is_unaligned = read_pointer & 0b11 != 0;
+
+            read_error = if bytes_to_read == 2 || is_unaligned {
+                target_core
+                    .core
+                    .read_word_16(read_pointer)
+                    .inspect(|val| {
+                        code_buffer_le.extend_from_slice(&val.to_le_bytes());
+                        read_pointer += 2;
+                    })
+                    .err()
+            } else if bytes_to_read == 4 {
+                target_core
+                    .core
+                    .read_word_32(read_pointer)
+                    .inspect(|val| {
+                        code_buffer_le.extend_from_slice(&val.to_le_bytes());
+                        read_pointer += 4;
+                    })
+                    .err()
+            } else {
+                // All supported architectures have either 16 or 32 bit instructions.
+                unreachable!();
```

Comment:
Can we return an error here? We already return an error from this function and it would be nice to return an error to the user instead of crashing if they manage to load the wrong arch ;)

Reply:
Fixed.